pkg/fuzzy/googletranslate.go

```diff
@@ -0,0 +1,137 @@
+	"golang.org/x/text/language"
+	"google.golang.org/api/option"
+	htransport "google.golang.org/api/transport/http"
+)
+
+// --------------------Definitions--------------------
+
+// Interface that defines some of the methods of the Google Translate client.
+// This interface helps to mock the Google Translate client in unit tests.
+// https://pkg.go.dev/cloud.google.com/go/translate#Client
+type googleClient interface {
+	Translate(ctx context.Context, inputs []string, target language.Tag, opts *translate.Options) ([]translate.Translation, error) //nolint:lll
+	io.Closer
+}
+
+// googleTranslate implements the TranslationService interface.
+type googleTranslate struct {
+	client googleClient
+}
+
+type GoogleTranslateOption func(*googleTranslate) error
+
+// WithClient sets the Google Translate client.
+func WithClient(c googleClient) GoogleTranslateOption {
+	return func(g *googleTranslate) error {
+		g.client = c
+		return nil
+	}
+}
+
+// WithDefaultClient creates a new Google Translate client with the API key from the viper.
+func WithDefaultClient(ctx context.Context) GoogleTranslateOption {
+	return func(g *googleTranslate) error {
+		var err error
+
+		apiKey := viper.GetString("translate_services.google_translate.api_key")
+		if apiKey == "" {
+			return fmt.Errorf("with default client: google translate api key is not set")
+		}
+
+		// Create new Google Cloud service transport with the base of OpenTelemetry HTTP transport.
+		trans, err := htransport.NewTransport(
+			ctx,
+			otelhttp.NewTransport(http.DefaultTransport),
+			option.WithAPIKey(apiKey),
+		)
+		if err != nil {
+			return fmt.Errorf("with default client: new transport: %w", err)
+		}
+
+		g.client, err = translate.NewClient(ctx, option.WithHTTPClient(&http.Client{Transport: trans}))
+		if err != nil {
+			return fmt.Errorf("with default client: new google translate client: %w", err)
+		}
+
+		return nil
+	}
+}
+
+// NewGoogleTranslate creates a new Google Translate service.
+func NewGoogleTranslate(
+	ctx context.Context,
+	opts ...GoogleTranslateOption,
+) (gt *googleTranslate, closer func() error, err error) {
+	gt = &googleTranslate{}
+
+	for _, opt := range opts {
+		if optErr := opt(gt); optErr != nil {
+			return nil, nil, fmt.Errorf("apply opt: %w", optErr)
+		}
+	}
+
+	// Ping the Google Translate API to ensure that the client is working.
+	_, err = gt.client.Translate(ctx, []string{"Hello World!"}, language.Latvian, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("google translate client: ping google translate: %w", err)
+	}
+
+	return gt, gt.client.Close, nil
+}
+
+// --------------------Methods--------------------
+
+func (g *googleTranslate) Translate(
+	ctx context.Context,
+	messages *model.Messages,
+	targetLang language.Tag,
+) (*model.Messages, error) {
+	if err := validateTranslate(messages, targetLang); err != nil {
+		return nil, fmt.Errorf("google translate: validate translate request: %w", err)
+	}
+
+	translatedMessages := model.Messages{
```

Comment:
Move this down closer to consumption! Always make sure it is efficient!

You declare a variable and reserve memory here, but if the translation fails, it was a wasted effort.